upgrading/upgrade-runnning-applications/4_migrate_contract.md

```diff
@@ -0,0 +1,138 @@
+
+rollbackComment : V2.Forum.Comment -> ContractId V1.Forum.Post -> V1.Forum.Comment
+rollbackComment V2.Forum.Comment{..} oldPostCid
+  = V1.Forum.Comment with commenter = rollbackUser commenter, post = oldPostCid, ..
+
+-- The actual MigrateUser contract describes what is the procedure and authorization to upgrade
+-- existing contracts.
+template MigrateUser
+  with
+    username: Party
+  where
+    signatory username
+    key username: Party
+    maintainer key
+
+    -- we want a non-consuming choice such that the we can migrate/rollback more contracts.
+    nonconsuming choice DoMigrateUser : ContractId V2.User.User
+      with
+        -- we choose a contract of V1 ...
+        userCid : ContractId V1.User.User
+      controller username
+      do
+        user <- fetch userCid
+        archive userCid
+        -- ... and upgrade it to V2
+        create $ upgradeUser user
+
+    -- this is the same as DoMigrateUser but with V1/V2 flipped.
+    nonconsuming choice DoRollbackUser : ContractId V1.User.User
+      with
+        userCid : ContractId V2.User.User
+      controller username
+      do
+        user <- fetch userCid
+        archive userCid
+        create $ rollbackUser user
+
+    -- analogous to the previous choices
+    nonconsuming choice DoMigratePost : ContractId V2.Forum.Post
+      with
+        postCid : ContractId V1.Forum.Post
+      controller username
+      do
+        V1.Forum.Post{..} <- fetch postCid
+        archive postCid
+        create $ V2.Forum.Post with user = upgradeUser user, ..
+
+    nonconsuming choice DoRollbackPost : ContractId V1.Forum.Post
+      with
+        postCid : ContractId V2.Forum.Post
+      controller username
+      do
+        V2.Forum.Post{..} <- fetch postCid
+        archive postCid
+        create $ V1.Forum.Post with user = rollbackUser user, ..
+
+    nonconsuming choice DoMigrateComment : ContractId V2.Forum.Comment
+      with
+        commentCid : ContractId V1.Forum.Comment
+        newPostCid : ContractId V2.Forum.Post
+      controller username
+      do
+        comment <- fetch commentCid
+        archive commentCid
+        create $ upgradeComment comment newPostCid
+
+    nonconsuming choice DoRollbackComment : ContractId V1.Forum.Comment
+      with
+        commentCid : ContractId V2.Forum.Comment
+        oldPostCid : ContractId V1.Forum.Post
+      controller username
+      do
+        comment <- fetch commentCid
+        archive commentCid
+        create $ rollbackComment comment oldPostCid
+</pre>
+
+1. The `MigrateUser` template offers every user of the mini social network of the
+   create-daml-app-0.1.0 package to migrate his `User`, `Post` and `Comment` contracts to the
+   updated version of the packages.
+1. We use qualified imports to avoid name clashes of the two imported modules. Note how we can
+   import the two equally named modules of the two packages with the given prefix we added in the
+   `daml.yaml` file in the previous step.
```

Comment:
```suggestion
   `../migration/daml.yaml`{{open}} file in the previous step.
```
Just wanna give the user a shortcut so s/he can open the file and take a look again.

Reply:
changed.